pennylane_lightning_kokkos/src/simulator/GateFunctors.hpp

```diff
@@ -196,85 +196,105 @@ template <class Precision, bool inverse = false> struct multiQubitOpFunctor {
 
     using KokkosComplexVector = Kokkos::View<Kokkos::complex<Precision> *>;
     using KokkosIntVector = Kokkos::View<std::size_t *>;
+    using ScratchViewComplex =
+        Kokkos::View<Kokkos::complex<Precision> *,
+                     Kokkos::DefaultExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using ScratchViewSizeT =
+        Kokkos::View<std::size_t *,
+                     Kokkos::DefaultExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using MemberType = Kokkos::TeamPolicy<>::member_type;
 
     KokkosComplexVector arr;
     KokkosComplexVector matrix;
-    KokkosIntVector indices;
     KokkosIntVector wires;
-    KokkosComplexVector coeffs_in;
     std::size_t dim;
     std::size_t num_qubits;
 
     multiQubitOpFunctor(KokkosComplexVector &arr_, std::size_t num_qubits_,
                         const KokkosComplexVector &matrix_,
                         KokkosIntVector &wires_) {
         dim = 1U << wires_.size();
-        indices = KokkosIntVector("indices", dim);
-        coeffs_in = KokkosComplexVector("coeffs_in", dim);
         num_qubits = num_qubits_;
         wires = wires_;
         arr = arr_;
         matrix = matrix_;
     }
 
     KOKKOS_INLINE_FUNCTION
-    void operator()(std::size_t kp) const {
-        const std::size_t k = kp * dim;
-        using Pennylane::Lightning_Kokkos::Util::bitswap;
+    void operator()(const MemberType &teamMember) const {
+        const std::size_t k = teamMember.league_rank() * dim;
+        ScratchViewComplex coeffs_in(teamMember.team_scratch(0), dim);
+        ScratchViewSizeT indices(teamMember.team_scratch(0), dim);
         if constexpr (inverse) {
 
-            for (size_t inner_idx = 0; inner_idx < dim; inner_idx++) {
-                std::size_t idx = k | inner_idx;
-                const std::size_t n_wires = dim;
-
-                for (std::size_t pos = 0; pos < n_wires; pos++) {
-                    size_t x = ((idx >> (n_wires - pos - 1)) ^
-                                (idx >> (num_qubits - wires[pos] - 1))) &
-                               1U;
-                    idx = idx ^ ((x << (n_wires - pos - 1)) |
-                                 (x << (num_qubits - wires[pos] - 1)));
-                }
-
-                indices[inner_idx] = idx;
-                coeffs_in[inner_idx] = arr[idx];
-            }
-
-            for (size_t i = 0; i < dim; i++) {
-                const auto idx = indices[i];
-                arr[idx] = 0.0;
-
-                for (size_t j = 0; j < dim; j++) {
-                    const std::size_t base_idx = j * dim;
-                    arr[idx] +=
-                        Kokkos::conj(matrix[base_idx + i]) * coeffs_in[j];
-                }
+            if (teamMember.team_rank() == 0) {
+                Kokkos::parallel_for(
```

Comment:
Can we use functor here instead of lambda function? I remember that Trevor mentioned that lambda function is not friendly for HIP.

Reply:
I like to use functors when there is a setup phase or the function needs be reused or tested, and lambdas otherwise. You bring a good point, but there have been a lot of bug fixes and improvements on the HIP backend since then. In any case, I'll pass those lambda into the `KOKKOS_LAMBDA` macro, which will transform them to functors. Is that ok?

Reply:
Not necessarily functor or `KOKKOS_LAMABDA`.  My only concern is whether it can be built with a HIP backend or not. PR #[21](https://github.com/PennyLaneAI/pennylane-lightning-kokkos/pull/21) could be a reference to build against the HIP backend.